### DISCUSSION TITLE
Getting past LazyDatabase in migrations and/or custom statements...

### DISCUSSION THREAD

**Discussion by JanezStupar** [simolus3/drift#2737] [Q&A] [answered]
So my project `drift_crdt` relies on a custom executor. Which does stuff to database queries under the hood.

In this particular example it filters out all the soft deleted columns by default.

Maybe this is a bad idea, but it seemed like a good idea at the time - because it allows an adopting app to continue working like it has before.

However there are cases when one needs to query deleted records too. I devised a way to accomplish that, but predictably I ran into issues.

So in this particular case I am having issues with migrations.

My problem is that during the migration, the `CrdtQueryExecutor` is not available yet. And I cannot access the mechanism I devised from the `LazyDatabase` instance that is available during the migration phase.

Below you can see the mechanism I hacked together to allow me to control the query behavior.
```dart
/// Allows access to the deleted records using the Drift API
/// [db] the database executor to query
/// callback the callback to execute, works with transactions too.
Future<R> queryDeleted<R>(CrdtQueryExecutor db, DelegateCallback<R> callback) async {
  (db.delegate as _CrdtDelegate).queryDeleted = true;
  ((db.delegate as _CrdtDelegate).transactionDelegate
          as _CrdtTransactionDelegate)
      .queryDeleted = true;
  final result = await callback();
  (db.delegate as _CrdtDelegate).queryDeleted = false;
  ((db.delegate as _CrdtDelegate).transactionDelegate
          as _CrdtTransactionDelegate)
      .queryDeleted = false;
  return result;
}
```

An example of usage of the `queryDeleted` wrapper:
```dart
    final result = await queryDeleted((db.executor) as CrdtQueryExecutor,
        () async => db.select(db.users).get());
```

I noticed that queries/statements written in migrations resolve to the actual database instance. But that doesn't do much for me...

I also noticed that I can just avoid the LazyDatabase - because I do not need to rely on an asynchronous API for the database location in my particular use case, but that is not something I can rely on in the long run and would cause issues to other potential users of the library.

So now I need a way to either access my delegated `CrdtQueryExecutor` in migrations OR maybe - which would be even more elegant, I could remake my wrapper into a custom operator/statement that could be used in the context of Drift queries.

Or perhaps there is something else I can do but have failed to notice thus far.

**Comment by simolus3** [comment, marked as answer]
At the moment, there is no reliable way to reach the underlying database. With a `LazyDatabase` we might make this possible by exposing the underlying field (not a big fan of that though), but we also support databases transparently running in a different isolate, and with those it gets much more complicated.

A crude workaround might be to run special, otherwise invalid, statements that are intercepted in the `CrdtQueryExecutor` to invoke the desired behavior:

```dart
Future<R> queryDeleted<R>(DelegateCallback<R> callback) async {
  await customStatement('query deleted tables as well');
  final result = await callback();
  await customStatement('stop querying deleted tables');
  return result;
}
```

(likely not those statements, but something like that).

Perhaps we can introduce an additional asynchronous channel to talk to specific databases, but at the moment I have no good overview of possible use-cases for that.

**Reply by JanezStupar** [reply]
I think your suggestion while imperfect is actually better than what I did originally.

Ideally a custom statement would be the best approach, but that is neither here nor there.
